File: backend/api/app/domain/repositories/sketch.py
from abc import ABC, abstractmethod
from typing import Iterable
from app.domain.entities.sketch import SketchEntity, Sketch
from app.domain.repositories import ContextManagerRepository
from core.common.either import Either
from core.errors.failure import Failure


class BaseWriteOnlyRepository(ContextManagerRepository):
    @abstractmethod
    async def create_sketch(self, sketch: Sketch, team_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        ...

    @abstractmethod
    async def update_sketch(self, sketch: Sketch, sketch_id: str, team_id: str,  user_id: str) -> Either[Failure, SketchEntity]:
        ...

    @abstractmethod
    async def delete_sketch(self, sketch_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        ...


class BaseReadOnlyRepository(ABC):
    @abstractmethod
    async def views_sketch(self, team_id: str, user_id: str) -> Either[Failure, Iterable[SketchEntity]]:
        ...

    @abstractmethod
    async def view_sketch(self, sketch_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        ...

class BaseRepository(BaseReadOnlyRepository, BaseWriteOnlyRepository, ABC):
    ...

File: mobile/lib/features/architect/domains/repositories/sketch.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failure.dart';
import '../entities/sketch.dart';

abstract class SketchRepository {
  Future<Either<Failure, Sketch>> create({
    required String title,
    required String teamId,
  });
  Future<Either<Failure, Sketch>> update({
    required String sketchId,
    required String title,
    required String teamId,
  });

  Future<Either<Failure, List<Sketch>>> views({
    required String teamId,
  });
  Future<Either<Failure, Sketch>> view(String sketchId);
  Future<Either<Failure, Sketch>> delete(String sketchId);
}
